### DISCUSSION TITLE
Large packets get fragments lost when proxied from RADSEC to RADIUS

### DISCUSSION THREAD

**Discussion by ghost** [radsecproxy/radsecproxy#97] [Q&A]
Large packets arriving from TLS side fragmented get only first fragment forwarded to RADIUS side.
[radsec.pcap.zip](https://github.com/radsecproxy/radsecproxy/files/6507330/radsec.pcap.zip)

Frame 10,11 arrive from RADSEC, and only one fragment is sent to RADIUS in frame 13

**Comment by ghost** [comment]
Corresponding log entries
May 19 08:33:56 2021: clientradputtls: Sent 462 bytes, Radius packet of length 462 to TLS peer RADSECServer
May 19 08:33:56 2021: radtlsget: got 1627 bytes
May 19 08:33:56 2021: buf2radmsg: message auth ok
May 19 08:33:56 2021: got Access-Challenge message with id 1
May 19 08:33:56 2021: replyh: passing Access-Challenge (id 9) to client RADIUSClients 
May 19 08:33:56 2021: signalling server writer
May 19 08:33:56 2021: udp server writer, got signal
May 19 08:33:56 2021: udpserverwr: refcount 3
May 19 08:33:56 2021: freerq: called with refcount 3
May 19 08:33:56 2021: udp server writer, waiting for signal
May 19 08:33:56 2021: freerq: called with refcount 2

**Comment by fmauchle** [comment]
Packet fragmentation is handled by the (linux)kernel not by radsecproxy itself. The fact that you see the first fragment means that radsecproxy has handed over the bytes to the kernel to be sent (again, it can only send all bytes or none at all - there is no code in radsecproxy to split a packet).

So your issue is most likely on the host itself or the network (depends on where you did the packet capture). From personal experience, missing fragments have always been caused by firewalls denying fragments.
Also note that if you use a capture filter e.g. for udp port 1812 with tcpdump (or wireshark), it won't capture the udp fragments beyond the first packet.

**Comment by CalvinVoxox** [comment]
To capture IP fragments, you will need to filter for the More Fragments
flag and Fragment Offset value in the IP header:

Matching fragmentation ?
> ------------------------
> - Matching MF (more fragment set) ? This would match the fragmented
> datagrams but wouldn't match the last
>   fragment (which has the 2nd bit set to 0).
> # tcpdump -i eth1 'ip[6] = 32'
> The last fragment have the first 3 bits set to 0... but has data in the
> fragment offset field.
> - Matching the fragments and the last fragments
> # tcpdump -i eth1 '((ip[6:2] > 0) and (not ip[6] = 64))'
> A bit of explanations :
> "ip[6:2] > 0" would return anything with a value of at least 1
> We don't want datagrams with the DF bit set though.. the reason of the
> "not ip[6] = 64"


https://github.com/SergK/cheatsheat-tcpdump/blob/master/tcpdump_advanced_filters.txt

In your case, you want to add the above to your previous filter. Something
like: '... or (ip[6:2] > 0)'. A few of the packets in your capture have the
Don't Fragment bit set so you don't want to ignore them.

Some firewalls have an option to simply block IP fragments or not, other
firewalls might have DoS protection that could trigger on IP fragments if
there are many of them.



Calvin Ellison

Systems Architect

***@***.***

+1 (213) 285-0555

<http://voxox.com>

<https://www.facebook.com/VOXOX/> <https://www.instagram.com/voxoxofficial/>
<https://www.linkedin.com/company/3573541/admin/>
<https://twitter.com/Voxox>

The information contained herein is confidential and privileged information
or work product intended only for the individual or entity to whom it is
addressed. Any unauthorized use, distribution or copying of this
communication is strictly prohibited. If you have received this
communication in error, please notify me immediately.
